api/src/main/java/io/grpc/CallOptions.java

```diff
@@ -28,55 +29,129 @@
 import java.util.concurrent.TimeUnit;
 import javax.annotation.CheckReturnValue;
 import javax.annotation.Nullable;
-import javax.annotation.concurrent.Immutable;
 
 /**
  * The collection of runtime options for a new RPC call.
  *
  * <p>A field that is not set is {@code null}.
  */
-@Immutable
 @CheckReturnValue
 public final class CallOptions {
   /**
    * A blank {@code CallOptions} that all fields are not set.
    */
-  public static final CallOptions DEFAULT = new CallOptions();
+  public static final CallOptions DEFAULT = new Builder().build();
 
-  // Although {@code CallOptions} is immutable, its fields are not final, so that we can initialize
-  // them outside of constructor. Otherwise the constructor will have a potentially long list of
-  // unnamed arguments, which is undesirable.
   @Nullable
-  private Deadline deadline;
-  
+  private final Deadline deadline;
+
   @Nullable
-  private Executor executor;
+  private final Executor executor;
 
   @Nullable
-  private String authority;
+  private final String authority;
 
   @Nullable
-  private CallCredentials credentials;
+  private final CallCredentials credentials;
 
   @Nullable
-  private String compressorName;
+  private final String compressorName;
 
-  private Object[][] customOptions;
+  private final Object[][] customOptions;
 
-  // Unmodifiable list
-  private List<ClientStreamTracer.Factory> streamTracerFactories = Collections.emptyList();
+  private final List<ClientStreamTracer.Factory> streamTracerFactories;
 
   /**
    * Opposite to fail fast.
    */
   @Nullable
-  private Boolean waitForReady;
+  private final Boolean waitForReady;
 
   @Nullable
-  private Integer maxInboundMessageSize;
+  private final Integer maxInboundMessageSize;
   @Nullable
-  private Integer maxOutboundMessageSize;
+  private final Integer maxOutboundMessageSize;
+
+  private CallOptions(Builder builder) {
+    this.deadline = builder.deadline;
+    this.executor = builder.executor;
+    this.authority = builder.authority;
+    this.credentials = builder.credentials;
+    this.compressorName = builder.compressorName;
+    this.customOptions = builder.customOptions;
+    this.streamTracerFactories = builder.streamTracerFactories;
+    this.waitForReady = builder.waitForReady;
+    this.maxInboundMessageSize = builder.maxInboundMessageSize;
+    this.maxOutboundMessageSize = builder.maxOutboundMessageSize;
+  }
+
+  private static class Builder {
```

Comment:
Just assign to the fields directly, as they are accessible now (since you did (2) and (3)).

```java
  public CallOptions withDeadline(@Nullable Deadline deadline) {
    Builder b = toBuilder(this);
    b.deadline = deadline;
    return b.build();
  }
```